documentation/specs/src/interoperability/ethereum-bridge/ethereum_smart_contracts.md

```diff
@@ -1,27 +1,182 @@
 # Ethereum Smart Contracts
 
-The set of Ethereum contracts should perform the following functions:
-
-- Verify bridge header proofs from Namada so that Namada messages can
-  be submitted to the contract.
-- Verify and maintain evolving validator sets with corresponding stake
-  and public keys.
-- Emit log messages readable by Namada
-- Handle ICS20-style token transfer messages appropriately with escrow &
-  unescrow on the Ethereum side
-- Allow for message batching
-
-Furthermore, the Ethereum contracts will whitelist ETH and tokens that
-flow across the bridge as well as ensure limits on transfer volume per epoch.
-
-An Ethereum smart contract should perform the following steps to verify
-a proof from Namada:
-
-1. Check the epoch included in the proof.
-2. Look up the validator set corresponding to said epoch.
-3. Verify that the signatures included amount to at least 2 / 3 of the
-   total stake.
-4. Check the validity of each signature.
-
-If all the above verifications succeed, the contract may affect the
-appropriate state change, emit logs, etc.
\ No newline at end of file
+## Contracts
+
+There are five smart contracts that make up an Ethereum bridge deployment.
+
+- Proxy
+- Bridge
+- Governance
+- Vault
+- wNAM
+
+### Proxy
+
+The _Proxy_ contract serves as a dumb storage for holding the addresses of other
+contracts, specifically the _Governance_ contract, the _Vault_ contract and the
+current _Bridge_ contract. Once deployed, it is modifiable only by the
+_Governance_ contract, to update the address for which contract is the current
+_Bridge_ contract.
+
+The _Proxy_ contract is fixed forever once the bridge has been deployed.
```

Comment:
Actually, it could possibly be replaced with a hard fork. Hopefully never necessary